torchgeo/trainers/byol.py

```diff
@@ -0,0 +1,393 @@
+        """Randomapply forward method."""
+        return x if random.random() > self.p else self.augm(x)
+
+
+class EncoderWrapper(Module):
+    """Encoder wrapper joining model and projection head."""
+
+    def __init__(
+        self,
+        model: Module,
+        projection_size: int = 256,
+        hidden_size: int = 4096,
+        layer: Union[str, int] = -2,
+    ) -> None:
+        """Initializes EncoderWrapper.
+
+        Keyword Args:
+            model: Model to encode
+            projection_size: Size of fist layer of projector MLP
+            hidden_size: Size of hidden layer of projector
+            layer: Layer from model to project
+        """
+        super().__init__()  # type: ignore[no-untyped-call]
+
+        self.model = model
+        self.projection_size = projection_size
+        self.hidden_size = hidden_size
+        self.layer = layer
+
+        self._projector: Optional[Module] = None
+        self._projector_dim = 256
+        self._encoded = torch.empty(0)  # type: ignore[attr-defined]
+        self._register_hook()
+
+    @property
+    def projector(self) -> Module:
+        """Wrapper of projector head."""
+        if self._projector is None:
+            self._projector = mlp(
+                self._projector_dim, self.projection_size, self.hidden_size
+            )
+        return self._projector
+
+    # See: https://towardsdatascience.com/how-to-use-pytorch-hooks-5041d777f904
+
+    def _hook(self, _: Any, __: Any, output: Tensor) -> None:
+        """Projection hook."""
+        output = output.flatten(start_dim=1)
+        if self._projector_dim is None:
+            # If we haven't already, measure the output size
+            self._projector_dim = output.shape[-1]
+
+        # Project the output to get encodings
+        self._encoded = self.projector(output)
+
+    def _register_hook(self) -> None:
+        """Register hook for forward."""
+        if isinstance(self.layer, str):
+            layer = dict([*self.model.named_modules()])[self.layer]
+        else:
+            layer = list(self.model.children())[self.layer]
+
+        layer.register_forward_hook(self._hook)
+
+    def forward(self, x: Tensor) -> Tensor:
+        """Pass through the model, and collect 'encodings' from our forward hook!"""
+        _ = self.model(x)
+        return cast(Tensor, self._encoded)
+
+
+class BYOL(LightningModule):
+    """BYOL implementation.
+
+    BYOL contains two identical Encoder networks. The first is trained
+    as usual, and its weights are updated with each training batch. The
+    second (referred to as the “target” network) is updated using a running
+    average of the first Encoder’s weights.
+    Citation: Grill JB, Strub F, Altché F, Tallec C, Richemond PH, Buchatskaya E,
+    Doersch C, Pires BA, Guo ZD, Azar MG, Piot B. Bootstrap your own latent: A
+    new approach to self-supervised learning. arXiv preprint arXiv:2006.07733.
+    2020 Jun 13.
```

Comment:
It would be better to provide a link to the paper instead, that's what we do for datasets.